File: Data Structures & Algorithms/two-integer-sum-ii/submission-2.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left, right = 0, len(numbers) - 1
        while left < right:
            diff = numbers[left] + numbers[right]
            print(diff, left, right)
            if (diff == target):
                return [left+1,right+1]
            elif diff < target:
                left += 1
            else:
                right -= 1
                
        return []

File: Data Structures & Algorithms/permutation-string/submission-2.js
class Solution {
    /**
     * @param {string} s1
     * @param {string} s2
     * @return {boolean}
     */
    checkInclusion(s1, s2) {
        if (s1.length > s2.length) return false;

        let right = 0;
        let left = 0;
        let s1Counts = s1.split('').reduce((acc, curr) => {
            acc[curr] = curr in acc ? acc[curr] + 1 : 1;
            return acc;
        }, {});
        while (right < s2.length) {
            let s1Copy = {...s1Counts};
            while (s2[right] in s1Copy) {
                if (s1Copy[s2[right]] < 0) break;
                s1Copy[s2[right]]--;
                if (Object.values(s1Copy).every(val => val === 0)) return true;
                right++;
            }
            left++;
            right = left;

        }
        return false;
    }
}
